### DISCUSSION TITLE
Peering Manager <-> NetBox Integration

### DISCUSSION THREAD

**Discussion by jamesditrapani** [peering-manager/peering-manager#498] [Ideas]
Discussion around Netbox to Peering Manager integration, what we want to achieve and how we want to achieve it.

**Comment by gmazoyer** [comment]
NetBox and Peering Manager being two different sources of truth, we should as much as possible avoid duplicating data between these for the sake of people using both of these projects.

That said, NetBox and PM both help in documenting some aspects of networks. While NetBox holds IP address assignments, devices (and much more), PM  can leverage NetBox's data for BGP sessions tracking (BGP requires IP addresses right?).

We should gather people ideas about possible use cases and what they'd like to see as interactions between NetBox and PM. The first post of this discussion will be updated with accepted ideas and we will draft a proper feature request once enough ideas collected. We will then explore what are our best options to implement the feature (webhooks config, NetBox plugin, etc).

We already have some issues mentioning ideas such as: #255 and #263.

**Reply by jlixfeld** [reply]
NTC forked NetBox and spun off [Nautobot](https://github.com/nautobot/nautobot/) which builds on NetBox by adding extensive support and much more direct integration for all things automation.  I suspect that many folks who use Peering Manager are interested in automation from a holistic perspective.  If they are already using NetBox, they are likely to migrate to Nautobot, in time, to extend their ability to automate all of the things.  Conversely, those who use Peering Manager but have minimal, or no other automation in their network will probably choose Nautobot over NetBox once they see the power of automation that Peering Manager shows, and they understand the different paths that Netbox and Nautobot will take.

So, all this to say that maybe a thought is to look at supporting integration with Nautobot as well.  From what I gather, most of the Nautobot API endpoints are more or less structure the as the Netbox API, save some renaming of path elements and endpoints, but the changes seem consistent, so maybe it wouldn't be too difficult.  That, and Nautobot has native support for GraphQL which may be interesting, or not.

That would be my $0.02.

Otherwise, keep up the great work, and thank you for everything you've all done with Peering Manager.

**Reply by mngan** [reply]
My current (semi-automated) workflow for deploying new peering sessions is to use Netbox to allocate/assign prefixes and IP's for a direct peering session, then commit the interface and IP/prefix information into Netbox. At the same time, I create the session in PM from the same information. This could be a NB plugin and provide a single API/UI entry point to interact with both NB and PM when creating a direct session. For adding new IX's and IX sessions we still just use the UI since the PeeringDB integration makes things fairly quick and easy. 

There are a number of pieces of data related to a BGP or IXP session
1) BGP Group/IX
2) Local router interface and IP/prefix
3) Local AS information
4) Peer IP
5) Peer AS information
6) BGP settings (enabled, multihop, ttl, passwords etc..)

If you were to build a NB plugin, much of the data entered is just being passed through to PM, and a little is being inserted into or queried from NB. NB is mainly responsible for #2 and possibly #4. #4 can be derived if using a point to point connection, otherwise it has to be supplied by the user. But NB could be used to validate the input since it should be able to check if the IP is in the same prefix as the local interface (unless using multihop). #1, #3 and #5 could be retrieved from PM assuming that it exists. Otherwise they need to be created. #6 needs to be user supplied, although you could rely on the defaults in most cases. 

You could make a case to do it the other way around, since PM requires a majority of the data. This just leaves the IP/Prefix allocation. The way I do this is I query the pool of Prefixes/IP's available for a router. I use a combination of Prefix Roles, Tags and Sites in NB. This does put some onus on the operator to set up and maintain their NB prefixes and devices in a way that can uniquely relate prefixes to routers for specific purposes. PM would need to provide a flexible enough way to query for those prefixes. It will not "just work" out of the box. Alternatively, you could assign a pool of prefixes to routers in PM by looking them up in NB and recording them in PM. PM will run through the list of prefixes when allocating a new prefix to find one that can allocate the link net prefix. This is probably a much simpler system than classifying and tagging your prefixes in NB and having a query system in PM. It is probably more accessible, but potentially requires more care and feeding when adding more IP blocks or routers. 

Not everyone is going to want to try and allocate the IP and build the BGP session in one step. So if interfaces and IP's for direct peering sessions are created in NB before BGP sessions are created. The interfaces/IP's could be tagged for PM. Then on the PM side when you go to create a new BGP session, you can pull up a list of IP's/Interfaces for the router that haven't been assigned to BGP groups yet. This would make filling in the BGP session information a little faster and avoid re-entry of the information. I guess a similar thing can be done when adding a new IX connection. PM could lookup the interface and router based on information that is in NB to pre-populate those fields for the new IX connection. For a UI driven workflow, this could be useful. This can also warn the user when they are trying to override those pre-populated fields since they are straying from the SoT.

**Comment by tardoe** [comment]
Here are my 2c: tldr, keep SoT for things in the appropriate places and make weak references.

I would suggest using NB as a read-only source for various pieces that PM might want to reference. E.g. use the NB GraphQL API to build the list of potential devices that a session can be associated with. Similarly an IX could be associated to a set of Netbox circuits as well so that a given interface/circuit combination could be linked across the two SoTs so that I could, for example ask: "hey, I need to take down et-0/0/1 on edge1, which BGP peers do I have to drain?" 

NB of recent has added great support for plugins - if you needed to expose a particular integration surface via a new API path etc, you could very easily build a plugin to do that for you. This would then be a requirement for those wishing to integrate PM with NB. I suspect you'd be fine with the vanilla API as it is today anyway.

The sticking point could be around broken references - e.g. you keep the id of a NB device in PM - if that device is deleted on the NB side, you'll need to handle that on the PM side.
